app/src/main/java/ingsw/group1/findmyphone/cryptography/SMSCipher.java

```diff
@@ -0,0 +1,145 @@
+        StringBuilder encryptedStr = new StringBuilder();
+        for (int i = 0; i < strInt.length; i++) {
+            String s = Integer.toHexString(strInt[i] ^ keyInt[i % keyInt.length]);
+            String s2 = SMSCipher.addPadding(s, 2);
+            encryptedStr.append(s2);
+        }
+        return encryptedStr.toString();
+    }
+
+    /**
+     * Decrypts the string using xor encryption with the key.
+     *
+     * @param stringToDecipher The string to be decrypted.
+     * @param key              The key used for decryption.
+     * @return The Decrypted string.
+     */
+    public static String decrypt(String stringToDecipher, String key) {
+        StringBuilder decrypted = new StringBuilder();
+        int[] keyInt = SMSCipher.fromStringToAscii(key);
+        int[] intArr = new int[stringToDecipher.length()];
+        String[] strArr = new String[stringToDecipher.length() / 2];
+        for (int i = 0; i < strArr.length; i++) {
+            strArr[i] = stringToDecipher.substring(i * 2, (i * 2) + 2);
+        }
+        for (int i = 0; i < strArr.length; i++) {
+            int decimal = Integer.parseInt(strArr[i], 16);
+            intArr[i] = decimal;
+            String s = String.valueOf(intArr[i] ^ keyInt[i % keyInt.length]);
+            int n = Integer.valueOf(s, 10);
+            decrypted.append((char) n);
+        }
+        return decrypted.toString();
+    }
+
+    /**
+     * Converts an integer to a string and adds to it the padding character.
+     * If stringToPad.length is grater or equals to @length than it only converts the intToPadd into a string.
+     *
+     * @param stringToPad The String that need the padding.
+     * @param length      The wanted length for the string.
+     * @return String with the correct length.
+     */
+    private static String addPadding(String stringToPad, int length) {
+        final char PARSING_CHARACTER = '0';
+        while (stringToPad.length() < length) {
+            stringToPad = PARSING_CHARACTER + stringToPad;
+        }
+        return stringToPad;
+    }
+
+    /**
+     * Converts every characters of the string into integer of Ascii table and save their values in an array of integer.
+     *
+     * @param stringToBeConverted The string to be converted
+     * @return An array of integer containing in position i the value of the character i of the string
+     */
+    static int[] fromStringToAscii(String stringToBeConverted) {
+        int[] integerArray = new int[stringToBeConverted.length()];
+        for (int i = 0; i < stringToBeConverted.length(); i++) {
+            integerArray[i] = (int) stringToBeConverted.charAt(i);
+        }
+        return integerArray;
+    }
+
+
+    /**
+     * Sets the password that will be used as a key for encryption and decryption.
+     *
+     * @param newPassword The password used for encryption and decryption.
+     */
+    public void setPassword(String newPassword) {
+        this.password = newPassword;
```

Comment:
Why a setter? What if someone doesn't call it and tries to encrypt with the null password? If it's a precondition (which I still wouldn't recommend), then you should write it somewhere!

Reply:
This setter is here because this guy is a singleton, getInstance(password) doesn't look good to me, but we should check if the password has been set before encrypting.
I think that changing the password should be allowed, if the user, at a certain point, wants to ignore all responses to previous messages, he just has to change the password.
We should note that having a singleton cipher makes so that our application will have ONE password for EACH GROUP of users. Do we want that? Or do we want passwords to be set for each pair of users? The last case would require an object pool.

Reply:
I totally missed you comment Niccolò, the idea was to have a symmetric password for each user, and the user keeps only his own password, requiring to input the password of another user in order to contact him